corpus2alpino/annotators/enrich_lassy.py

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python3
+"""
+Allows enriching nodes of a Lassy annotation using a dictionary
+"""
+from typing import cast, Dict, List
+
+import csv
+from lxml import etree
+
+from corpus2alpino.abstracts import Annotator
+from corpus2alpino.annotators.alpino import ANNOTATION_KEY
+from corpus2alpino.log import LogSingleton
+from corpus2alpino.models import Document, Utterance
+
+
+class EnrichLassyAnnotator(Annotator):
+    def __init__(self, enrichment_file: str):
+        """Creates an annotator which will add additional attributes to
+        nodes when they match the properties specified in an enrichment
+        file.
+
+        Arguments:
+            enrichment_file {str} -- Path to a comma-separated enrichment file
+            matching attributes columns should be prefixed with @, assignments
+            without. For example the following CSV would match 
+            <node pos="noun" num="pl" /> and add penn_pos="NNS":
+
+            @pos,@num,penn_pos
+            noun,pl,NNS
+            noun,,NN
+            pron,pl,NNPS
+            pron,,NNP
+            det,,DT
+
+            Empty matchers are skipped, the first match is assigned.
+
+        Raises:
+            Exception: Exception is raised if it could not load or
+            parse the enrichment file
+        """
+        with open(enrichment_file) as enrichment:
+            dialect = csv.Sniffer().sniff(enrichment.read(1024))
+            enrichment.seek(0)
+            reader = csv.reader(enrichment, dialect)
+            matchers = {}
+            assigners = {}
+
+            headers = next(reader)
+            for i, header in enumerate(headers):
+                if header.startswith("@"):
+                    matchers[header[1:]] = i
+                else:
+                    assigners[header] = i
+
+            self.enrichments = cast(List[Enrichment], [])
+
+            for row in reader:
+                for i, value in enumerate(row):
+                    self.enrichments.append(Enrichment(
+                        {
+                            key: row[index] for key, index in matchers.items() if row[index]
+                        },
+                        {
+                            key: row[index] for key, index in assigners.items() if row[index]
+                        }))
```

Comment:
Personally, I really like making the language do as much work as possible, like you do in these dictionary comprehensions. However, from recent experiences in readit, I learned that other developers aren't always as fluent in the language and that they might even give up on reviewing your code when they run into a notation that looks alien to them.

Solution: just keep using such notations, but put a comment above them that (1) gives the name of the construct so less eloquent colleagues can look it up ("dictionary comprehension" in this case) and (2) summarizes what the snippet is doing.